---
title: "Sample and reference filtering"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.retina = 2)
library(tidyverse)
library(ggpubr)
library(ggtree)
library(phytools)
library(ggrepel)
source("scripts/my_color.R")
```

## Sample filtering

### 1. Blast analysis of mtDNA
We mapped reads to A.tenius mitochondrial DNA sequence and used `ANGSD doFasta2` to get the most like mitochondrial sequences of every sample. We then blast these mitochondrial genome to NCBI nt database to check to best match species. We found 10 samples with a best hit to other Acropora spp.

```bash
angsd -dofasta 2 -doCounts 1 -i {sample}_mito.bam -out fasta/{sample}_mito
gzip -d fasta/{sample}_mito.fa.gz
blastn -task megablast -db nt -query {sample}_mito.fa -remote -outfmt  '6 qseqid sseqid pident mismatch gapopen evalue staxids sscinames scomnames sskingdoms stitle' -out {sample}.blast
```

Next, we build a tree using iqtree with these sequences. There are 9 samples from Arlington reef and 1 from John Brewer Reef form a separate cluster from other samples. As we included mitogenome of *A.tenuis* and *A.echinata*, this is consistent with our blast results as 8 ARL samples and 1 JB samples matched with *A.echinata* and 1 ARL sample matched to *A.florida*.

```{r, fig.height=6.8, fig.width=5}
treefile <- "data/hpc/qc/sample_filtering/mitogenome/all_mito.aln.fasta.treefile"
tree <-read.tree(treefile)
tree<- midpoint.root(tree) 
df <- data.frame(label=tree$tip.label, pop=map_chr(tree$tip.label %>% str_split("_|-"), first))
tree <- full_join(tree,df,by='label') 

ggtree(tree,color="firebrick") + geom_tiplab(aes(color=pop),size=1.1) + scale_color_manual(values = site_colors(), guide=F)
```

We thus excluded these 10 samples for the following analysis.

### Identification of relatives

As clones or close related individuals could confound the results, we used `ngsRelate` to calculate the pairwise relatedness of all samples and remove one from any related sample pairs. We used glf format (-doGlf 3), however, vcf files with PL tage also can be used.

```bash
zcat atenius.mafs.gz | cut -f 5 |sed '1d' < freq
ngsRelate -g atenius.glf.gz -n 228 -f freq -p 30 -l 0.05 -O atenius.ngsrelate.txt
```

We then plotted the combinations of R0,R1 and R1-KING-robust kinship statistics of each pair of samples to infer relationships follow the methods in https://onlinelibrary.wiley.com/doi/10.1111/mec.14954.

There are three pairwise values clear deviated from the main cluster and are potential FS/OF relationship. These outliers are from Magnetic Island , we thus kept one with higher coverage from each pair in the following analysis. They have KING-robust kinship > 0.397, R1>1.94, R0<0.000016.

 - 117-119 (MI-2-21_S5, MI-2-24_S36)
 - 111-123 (MI-1-6_S37, MI-2-3_S13)
 - 103-127 (MI-1-16_S8, MI-2-9_S25)
 
 samples to remove (the one with lower sequencing depth from each pair):
 
 - MI-2-21_S5
 - MI-2-3_S13
 - MI-1-16_S8

```{r fig.show="hold", out.width="50%"}
excluded_samples <- c("ARL_13_S67","ARL_14_S68","ARL_1_S61","ARL_20_S73","ARL_21_S74","ARL_22_S75","ARL_23_S76","ARL_3_S63","ARL_5_S65","JB_5_S5")

samples <- read_tsv("data/hpc/qc/sample_filtering/relatedness/sample.list",col_names = F) %>% select(X1) %>% pull
sample_ids <- seq(0,227)
names(sample_ids) <- samples

res <- read_tsv("data/hpc/qc/sample_filtering/relatedness/atenius.ngsrelate.txt") %>% 
  select(a,b,R0,R1,KING) %>% mutate(a_name=names(sample_ids[a+1]),b_name=names(sample_ids[b+1]), type=ifelse(substr(a_name,1,3)==substr(b_name,1,3),"within","between"),name=paste(a_name,b_name,sep = " - ")) %>% filter(!a_name %in% excluded_samples,! b_name %in% excluded_samples)

ggplot(res,aes(x=R1,y=R0,label=name)) + geom_point(color="blue3",size=1)+
  geom_point(data=res %>% filter(R0<0.00002),size=1,color="red") +
  geom_text_repel(data=res %>% filter(R0<0.00002), size=2.5) + theme_pubr()

ggplot(res,aes(x=R1,y=KING,label=name)) + geom_point(color="blue3",size=1) + 
  geom_point(data=res %>% filter(R0<0.00002),size=1,color="red") +
  geom_text_repel(data=res %>% filter(R0<0.00002), size=2.5) + theme_pubr()
```

After this, we left 215 samples to go.

## Reference filtering

### 1. Mappability

We excluded genomic regions that can not be uniquely mapped by estimate the mappability of each site across the genome using [genmap v1.3.0](https://github.com/cpockrandt/genmap) with 50bp kmer size.

```bash
genmap -F reference fasta -I genmap_index
genmap map -K 50 -E 2 -I genmap_index -O reference_mappability_K50_E2 -t -w -bg
awk '$4==1{print $1"\t"$2"\t"$3}' reference_mappability_K50_E2.bedgraph > reference_mappability_K50_E2.unique.bed
```

There were 264,895,970 bp left.


### 2.Simple repeat regions

We used mdust to remove genomic regions that are short simply repeats. Next, we subtract these regions from mappability region in following analysis.

```bash
mdust reference.fasta -c |cut -f1,3,4 > reference.mdust.bed
bedtools subtract -a reference_mappability_K50_E2.unique.bed -b reference.mdust.bed
```

After this, 261,651,886 bp left.

### 3.Global depth

Genome regions with a lot of reads mapped or very little reads mapped should also be excluded in our analysis to avoid bias. We used `ANGSD` to calculate mapping depth of all samples across those genomic regions.

```bash
angsd sites index reference_mappability_K50_E2.unique.bed
angsd -minMapQ 30 -minQ 30 -doCounts 1 -doDepth 1 -dumpCounts 1 -maxdepth 3000 -b ind215_bam.list -out atenius  -nThreads 20 -sites reference_mappability_K50_E2.unique.bed
```

```{r}
g_depth <- read_table2("data/hpc/qc/reference_filtering/depth/atenius.depthGlobal", col_names = F, col_types = cols("X3002"=col_skip()))
g_depth_cumsum<-cumsum(as.numeric(g_depth))
q95 <- g_depth_cumsum[length(g_depth_cumsum)]*0.95
q05 <- g_depth_cumsum[length(g_depth_cumsum)]*0.05
q95.threshold <- min(which(g_depth_cumsum > q95))
q05.threshold <- min(which(g_depth_cumsum > q05))
```
