---
title: "Quality control: sample and reference filtering"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE,fig.retina = 2)
library(tidyverse)
library(ggpubr)
library(ggtree)
library(phytools)
library(ggrepel)
source("scripts/my_color.R")
```

## Sample filtering

Before proceeding with our main population genetic analyses we wanted to ensure that (a) all samples were from the correct species *A. kenti* and (b) that none were clones or close kin.  This filtering was performed as follows, and resulted in removal of 10 misidentified samples and 6 close kin, leaving 212 samples remaining. 

### 1. Identification of samples that are not *A. kenti* via mtDNA

We mapped reads to the *A. kenti* mitochondrial genome and used `ANGSD doFasta2` to get the most likely mitochondrial sequences of every sample. We then blasted these mitochondrial genome sequences against the NCBI nt database to check to best match species. We found 10 samples with a best hit to other Acropora spp.

```bash
angsd -dofasta 2 -doCounts 1 -i {sample}_mito.bam -out fasta/{sample}_mito
gzip -d fasta/{sample}_mito.fa.gz
blastn -task megablast -db nt -query {sample}_mito.fa -remote -outfmt  '6 qseqid sseqid pident mismatch gapopen evalue staxids sscinames scomnames sskingdoms stitle' -out {sample}.blast
```

Next, we combined the mito genome sequences of our samples with the reference mito genome of *A. echinata* as this species was identified via blast as a match to some samples.  We then aligned sequences using `mafft` and built a tree with iqtree (v1.6.4). 

```bash
mafft --auto --thread 10 all_mito.fasta > all_mito.aln.fasta
iqtree -s all_mito.aln.fasta
```

This revealed 9 samples from Arlington reef and 1 from John Brewer Reef that form a separate cluster from other samples along with *A. echinata*. This is consistent with blast results where these same samples matched either *A. echinata* or *A. florida* (1 sample), and suggests that they were misidentified during sample collection.

```{r}
excl <- read_tsv("data/qc/excluded_samples.txt",col_names = "id")

treefile <- "data/qc/all_mito.aln.fasta.treefile"
tree <-read.tree(treefile)
tree<- midpoint.root(tree) 

excl_anc <- ape::getMRCA(tree,excl$id[1:10])

df <- data.frame(label=tree$tip.label, pop=map_chr(tree$tip.label %>% str_split("_|-"), first)) %>% 
  mutate(species_label = case_when(
    label=="LC201841.1" ~ "A. echinata (LC201841.1)",
    label=="AF338425.1" ~ "A. kenti (AF338425.1)",
    .default = ""
  ))
tree <- full_join(tree,df,by='label')

ggtree(tree,color="firebrick",layout = "fan") + 
  geom_tippoint(aes(color=pop),size=1) +
  geom_tiplab(aes(label=species_label),size=2.5) +  
  scale_color_manual(values = site_colors()) + 
  geom_hilight(node=excl_anc, fill="steelblue", alpha=.6) +
  theme_tree() + 
  theme(legend.position = "left", legend.title = element_blank())
ggsave("figures/Figure_S2.png",width = 8,height = 6)
```

**Figure 1: Maximum likelihood tree of consensus mitochondrial genome sequences for all samples**. Tip labels are colored by reef location with the exception of reference sequences for *A. echinata* (LC201841.1) and *A. kenti* (AF338425.1) which are shown in grey. Samples within the highlighted clade (10 samples) were excluded from further analyses. 

### 2.Identification of clones and close kin

As clones or closely related individuals could confound the results, we used `ngsRelate` to calculate the pairwise relatedness and identify any clonal or family group clusters that may be present. 

To prepare inputs for `ngsRelate` we calculated genotype likelihoods with `ANGSD`, retaining only common polymorphic sites (MAF>0.05, SNP p_value <1e-6).

```bash
angsd -ref reference.fasta -b all_bam.list -GL 2 -doGlf 2 \
-doMajorMinor 1 -SNP_pval 1e-6 -doMaf 1 \
-minMapQ 30 -minQ 30 -minMaf 0.05 -out atenius -nThreads 30

zcat atenius.mafs.gz | cut -f 5 |sed '1d' > freq
ngsRelate -G atenius.beagle.gz -n 228 -f freq -p 30 -l 0.05 -O atenius.ngsrelate.txt
```

Following @{Waples2019-kn} we plotted the combinations of R1,R0 and R1,KING-robust kinship statistics of each pair of samples. This revealed eight outlying pairs that clearly deviated from the main cluster and that potentially represent samples that are close kin. All of these samples are from Magnetic Island. We then kept one with higher coverage (bold) from each pair for further analysis. They all have KING-robust kinship > 0.2998, R1>0.766, R0<0.005.

 - 117-119 (MI-2-21_S5, **MI-2-24_S36**)
 - 111-123 (**MI-1-6_S37**, MI-2-3_S13)
 - 103-127 (MI-1-16_S8, **MI-2-9_S25**)
 - 101-117 (MI-1-12_S2, MI-2-21_S5)
 - 109-115 (**MI-1-4_S10**, MI-2-16_S19)
 - 101-119 (MI-1-12_S2, **MI-2-24_S36**)
 - 101-124 (MI-1-12_S2, **MI-2-4_S26**)
 - 110-126 (MI-1-5_S35, **MI-2-8_S30**)
 
```{r}
excluded_samples <- c("ARL_13_S67","ARL_14_S68","ARL_1_S61","ARL_20_S73","ARL_21_S74","ARL_22_S75","ARL_23_S76","ARL_3_S63","ARL_5_S65","JB_5_S5")

samples <- read_tsv("data/qc/sample_ids228.txt",col_names = F) %>% select(X1) %>% pull
sample_ids <- seq(0,227)
names(sample_ids) <- samples

res <- read_tsv("data/qc/atenius.ngsrelate.txt") %>% 
  select(a,b,R0,R1,KING) %>% mutate(a_name=names(sample_ids[a+1]),b_name=names(sample_ids[b+1]), type=ifelse(substr(a_name,1,3)==substr(b_name,1,3),"within","between"),name=paste(a_name,b_name,sep = " - ")) %>% filter(!a_name %in% excluded_samples,! b_name %in% excluded_samples)

p1<-ggplot(res,aes(x=R1,y=R0,label=name)) + geom_point(color="blue3",size=1)+
  geom_point(data=res %>% filter(KING>0.25),size=1,color="red") +
  geom_text_repel(data=res %>% filter(KING>0.25), size=2.5) + theme_pubr(base_size = 12)

p2<-ggplot(res,aes(x=R1,y=KING,label=name)) + geom_point(color="blue3",size=1) + 
  geom_point(data=res %>% filter(KING>0.25),size=1,color="red") +
  geom_text_repel(data=res %>% filter(KING>0.25), size=2.5) + theme_pubr(base_size = 12)

cowplot::plot_grid(p1,p2,align = "h",ncol = 2,labels = c("(a)","(b)"),label_size = 12)
ggsave(filename = "figures/Figure_S3.png",dpi = 300,width = 8,height = 6)

#res %>% arrange(desc(KING)) %>% head(n=8) %>% select(a,b,name)
```

**Figure 2: Pairwise relatedness metrics for all pairs of samples**: R0, R1 and KING statistics calculated with `ngsRelate` for all 218 *A. tenuis* samples. Pairs shown in red represent likely close-kin relationships and were used to select 6 samples for removal.

## Reference filtering

To avoid spurious results due to SNPs called within duplicated and low complexity regions we identified a set of genomic regions where analyses could be perfomed reliably. The resulting intervals are provided as a bed file, [reference_mappability_K50_E2.unique_mdust.bed](data/qc/reference_mappability_K50_E2.unique_mdust1M_depth.bed) and were generated as follows;

### 1. Mappability

We excluded genomic regions that can not be uniquely mapped by estimating the mappability of each site across the genome using [genmap v1.3.0](https://github.com/cpockrandt/genmap) with 50bp kmer size.

```bash
genmap -F reference fasta -I genmap_index
genmap map -K 50 -E 2 -I genmap_index -O reference_mappability_K50_E2 -t -w -bg
awk '$4==1{print $1"\t"$2"\t"$3}' reference_mappability_K50_E2.bedgraph > reference_mappability_K50_E2.unique.bed
```

After this analysis approximately 267Mb of regions regions were left.

### 2.Simple repeat regions

We used mdust to remove genomic regions that are short simple repeats and subtracted these regions from the mappable regions identified in step 1.

```bash
mdust reference.fasta -c |cut -f1,3,4 > reference.mdust.bed
bedtools subtract -a reference_mappability_K50_E2.unique.bed -b reference.mdust.bed > reference_mappability_K50_E2.unique_mdust.bed
```

After this analysis approximately 263.8Mb of regions regions were left.

### 3. Short scaffolds

We also removed scaffolds with a length shorter than 1Mb. When combined with filtering steps 1 and 2 this resulted in 263,7Mb accessible bases.

## SNP filtering by depth

Genomic regions with extremely high or low coverage were also excluded because these can indicate problematic regions such as collapsed repeats (high cov) or falsely duplicated regions (low cov). We used `ANGSD` to calculate mapping depth of all samples across filtered genomic regions with minMapQ>30 and minQ >30. 

```bash
angsd sites index reference_mappability_K50_E2.unique_mdust.bed
angsd -minMapQ 30 -minQ 30 -doCounts 1 -doDepth 1 -dumpCounts 1 -maxdepth 3000 -b ind215_bam.list -out atenius  -nThreads 20 -sites reference_mappability_K50_E2.unique_mdust.bed
```

```{r,fig.width=6.2}
g_depth <- read_table2("data/qc/atenius.unique_mdust.ind212.depthGlobal", col_names = F, col_types = cols("X3002"=col_skip()))
g_depth_cumsum<-cumsum(as.numeric(g_depth))
q99 <- g_depth_cumsum[length(g_depth_cumsum)]*0.99
q01 <- g_depth_cumsum[length(g_depth_cumsum)]*0.01
q99.threshold <- min(which(g_depth_cumsum > q99))
q01.threshold <- min(which(g_depth_cumsum > q01))

plot(g_depth %>% t(),type="l",xlab = 'Global depth (count)', ylab = 'Number of sites')
abline(v = q99.threshold, col='red', lty=2)
abline(v = q01.threshold, col='red', lty=2)
legend('topright', legend = c(paste('1% thres:', q01.threshold, sep=''), paste('99% thres: ', q99.threshold, sep='')),  bty = "n")

```

This allowed us to identify minimum and maximum coverage thresholds as the 1% and 99% quantiles of global depth respectively. These values were used in ANGSD commands via the `-minDepth=17` and `-maxDepth=1102` filters to remove SNPs with depth values outside these bounds. The addition of these filters reduced the total number of accessible bases to 258.4Mb. 
