---
title: "Variant filteration and genotype likelihood"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.retina = 2)
library(tidyverse)
```

After getting the raw VCF files from gatk4 workflow, we applied some filtering. 

- We excluded sites with low quality score (GQ< 20); low quality by depth score (QD< 2); high Fisher strand score (FS> 60); low mapping quality (MQ< 30); low read position rank sum score (ReadPosRankSum< −4); high strand odds
ratio (SOR > 3); and low mapping quality rank sum score (MQRankSum < −12.5 & MQRankSum >12.5) using bcftools v1.12

- We removed sites within 10bp around indels

- We remove sites located in simple repetitive regions which are identified by mdust.

- We only included sites with average depth between 1-8.9X (We kept sites with mean depth between ~d/3 - d+3*sqrt(d), 1-8.9) using vcftool v0.1.16

- Eventually, we only used biallelic snps

```bash
bcftools index -t 10 gatk.vcf.gz
tabix -l gatk.vcf.gz > scaffolds.list
bcftools view -r {scaffold} -Oz -o {scaffold}.gatk.vcf.gz gatk.vcf.gz

mdust reference.fasta -c |cut -f 1,3,4 > reference.mdust.bed

bcftools filter -i 'FS<60.0 && SOR<3 && MQRankSum>-12.5 && MQRankSum<12.5 && ReadPosRankSum>-4.0 && QD>2.0 && MQ>30.0 && QUAL>30' --SnpGap 10 --threads 10 -Oz -o {scaffold}.gatk.hf_indel10bp.vcf.gz {scaffold}.gakt.vcf.gz

vcftools --gzvcf {scaffold}.gatk.hf_indel10bp.vcf.gz --recode --recode-INFO-all \
--minQ 30 --min-alleles 2 --max-alleles 2 \
--min-meanDP 1 --max-meanDP 8.9 \
--remove-indels --exclude-bed reference.mdust.bed \
--stdout | bgzip > {scaffold}.gatk.hf_indel10bp.gq_depth.vcf.gz

bcftools concat -Oz -o atenius.gatk.filtered.vcf.gz *.hf_indel10bp.gq_depth.vcf.gz
```

We didn't filtering by missing data at this stage for two reasons:

- Vcftools treats data as genotypes which we won't use since this is from lcWGS.

- During joint-genotype calling, we found gatk v4.2.5 we output homologous genotype 0/0 instead of ./. even at site with zero depth in sample. We tried to update to v4.2.6, the bug haven't resolved. However, if we downgraded to v4.1.19, it correctly annotated those missing genotype. 

Since most of our analyses are based on genotype likelihood (account for the genotype uncertainty) produced by angsd, we decided to filter sites with high number of missing data in angsd. (We did test whether it generate same results with gatk v4.2.5 and gatk v4.1.9 using data from several chromosomes/scaffolds, and the answer is positive).

PS: we didn't apply filtering based on deviation from hwe which is often used to remove genotyping error. However, the fact that we couldn't get accurate genotype form our lcWGR and many other factors can cause departures from HWE made us not do so.

**Convert vcf to beagle format**

We used angsd v0.937 to generate beagle format of our callset and filtering.
```bash
angsd -vcf-pl atenius.gatk.filtered.vcf.gz -out atenius -fai reference.fasta.fai \
    -nind 228 -doMaf 1 -doGlf 2 -doMajorMinor 1 -minmaf 0.01 -nThreads 48 -minInd 114
```

**LD pruning for PCA and Admixture analysis**

We generated a set of unlinked snps using ngsLD v1.1.1 and the perl script provided to keep unlinked sites.
```bash
zcat atenius.mafs.gz |sed '1d'|cut -f1,2 > snp_pos.txt
wc -l snp_pos.txt
ngsLD --geno atenius.beagle.gz --probs \  
    --pos snp_pos.txt --n_ind 228 --n_sites 4033852 \
    --min_maf 0.05 --n_threads 30 \
    --out atenius.beagle.ld

perl ngsLD/scripts/prune_graph.pl --in_file atenius.beagle.ld \
    --max_kb_dist 5 --min_weight 0.5 --out atenius.beagle.unlinked.id
```

