---
title: "Admixture analysis"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.retina = 2)
library(tidyverse)
library(ggpubr)
source("scripts/my_color.R")
sample_ids <- read_tsv("data/hpc/pcangsd/sample_ids.txt", col_names = "sample_id",show_col_types = FALSE) %>% pull(sample_id)
sample_table <- read_csv("data/hpc/qc/summary_data.csv",show_col_types = FALSE) %>% select(sample_id,pop,location,reef)
```

PCA anlaysis showed that all samples form three groups with one main group including most samples and one group composed of samples from magnetic island and a small group including several samples from Arlington reef and one from John Brewer Reef. Here, we infered the genetic admixture proportion of each individual. Same, we used genotype likelihood data that we used in PCA analysis, to keep it consistent, we also used whole-genome snps, LD pruned snps, and snps filtered by population specific missing data.

We used `NGSadmix` v32 to infer individual's ancestry while considering the uncertainty in genotypes with number of clusters setted to 2 and 3.

```bash
NGSadmix -likes atenius.beagle.gz -K 2 -outfiles atenius.K2 -P 20 -minMaf 0.05 -minInd 114
NGSadmix -likes atenius.beagle.gz -K 3 -outfiles atenius.K3 -P 20 -minMaf 0.05 -minInd 114
```

### 1. Admixture based on all SNPs

```{r, fig.height=14, fig.width=10}
q2_admixture <- read_table("data/hpc/admixture/atenius.K2.qopt", col_names = c("X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p1 <- ggplot(q2_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=2)") + 
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_blank())


q3_admixture <- read_table("data/hpc/admixture/atenius.K3.qopt", col_names = c("X3","X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p2 <- ggplot(q3_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=3)") + 
  scale_fill_brewer(palette = "Set3") + 
  scale_color_brewer(palette = "Set3") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_text(size=12))

cowplot::plot_grid(p1,p2,nrow=1)

```

### 2. Admixture based on sites with < 50% missing data in each population

see how we get this dataset in [PCA analysis](04.PCA_analysis.md)

```{r, fig.height=14, fig.width=10}
q2_admixture <- read_table("data/hpc/admixture/atenius.pop_miss05.K2.qopt", col_names = c("X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p1 <- ggplot(q2_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=2)") + 
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_blank())


q3_admixture <- read_table("data/hpc/admixture/atenius.pop_miss05.K3.qopt", col_names = c("X3","X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p2 <- ggplot(q3_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=3)") + 
  scale_fill_brewer(palette = "Set3") + 
  scale_color_brewer(palette = "Set3") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_text(size=12))

cowplot::plot_grid(p1,p2,nrow=1)

```

When using this dataset, it seems exaggerate/emphasis the divergence in Arlington reef subgroup (consistent with PCA), and the magnetic island seems to be hybrids at K=2. I think this is because samples from magnetic island all have lower genome mapping coverage, so eventually, we only kept ~500k snps that are shared in all populations. However, the divergence of magnetic island samples also come from those unmapped regions (could be too divergent to map?) and this was very different from others.

### 3. Admixture based on unlinked SNPs



```{r, fig.height=14, fig.width=10}
q2_admixture <- read_table("data/hpc/admixture/atenius.ld_pruned.K2.qopt", col_names = c("X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p1 <- ggplot(q2_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=2)") + 
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_blank())


q3_admixture <- read_table("data/hpc/admixture/atenius.ld_pruned.K3.qopt", col_names = c("X3","X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p2 <- ggplot(q3_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=3)") + 
  scale_fill_brewer(palette = "Set3") + 
  scale_color_brewer(palette = "Set3") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_text(size=12))

cowplot::plot_grid(p1,p2,nrow=1)
```

This gave similar results, when using unlinked snps, the separation seems to be stronger. Very unexpected.Dose this mean their genomes shared some large segements homozygousity while rest is diverging? 

What are those regions? Those keep them breeding? 

### Using recalibrated bam files in site restricted analysis

**1.popmiss05**

```{r, fig.height=14, fig.width=10}
q2_admixture <- read_table("data/hpc/admixture/atenius.recal.pop_miss05.K2.qopt", col_names = c("X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p1 <- ggplot(q2_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=2)") + 
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_blank())


q3_admixture <- read_table("data/hpc/admixture/atenius.recal.pop_miss05.K3.qopt", col_names = c("X3","X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p2 <- ggplot(q3_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=3)") + 
  scale_fill_brewer(palette = "Set3") + 
  scale_color_brewer(palette = "Set3") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_text(size=12))

cowplot::plot_grid(p1,p2,nrow=1)

```

**2.LD_pruned**

```{r, fig.height=14, fig.width=10}
q2_admixture <- read_table("data/hpc/admixture/atenius.recal.ld_pruned.K2.qopt", col_names = c("X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p1 <- ggplot(q2_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=2)") + 
  scale_fill_brewer(palette = "Set2") + 
  scale_color_brewer(palette = "Set2") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_blank())


q3_admixture <- read_table("data/hpc/admixture/atenius.recal.ld_pruned.K3.qopt", col_names = c("X3","X2","X1")) %>%
  add_column(sample_id = sample_ids) %>% 
  gather(cluster, proportion, -sample_id) %>% 
  left_join(sample_table) %>% 
  mutate(pop_order = site_order()[pop])

p2 <- ggplot(q3_admixture ,aes(x=reorder(sample_id,pop_order),y=proportion)) + 
  geom_bar(aes(fill=cluster,color=cluster),stat="identity") + 
  facet_grid(reorder(location,desc(pop_order))~., scales = "free_y") + coord_flip() + theme_pubclean() + labs(x="",y="Proportion (K=3)") + 
  scale_fill_brewer(palette = "Set3") + 
  scale_color_brewer(palette = "Set3") + 
  theme(axis.text.y = element_blank(),
        legend.position = "none", 
        axis.ticks = element_blank(),
        panel.border = element_blank(), 
        panel.background = element_blank(),
        panel.grid.minor = element_blank(),
        strip.text = element_text(size=12))

cowplot::plot_grid(p1,p2,nrow=1)
```