---
title: "Analysis of Inversion Genotypes"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.retina = 2)
```

```{r}
source("scripts/my_color.R")
library(tidyverse)
library(ggpubr)
```

Inversions give rise to a characteristic pattern of genetic structure when PCA is applied to SNPs within the inversion region. This structure groups samples into three clusters corresponding to three potential genotypes of the inversion itself, ie A/A, A/B, B/A where A and B represent the two orientations of the inversion.  In cases where the clusters are clear it is possible to infer the genotypes of samples according to their cluster membership and thereby investigate basic genetic properties of the inversion alleles.  

To assign samples to clusters we use kmeans clustering (with k=3) based on position along PC1 and with the initial centers of clusters at min(PC1), (min(PC1)+max(PC1))/2, max(PC1) for genotypes A/A, A/B and B/B respectively. 

```{r}
sample_ids <- read_tsv("data/hpc/inshore_offshore/north187_sample_id.txt", col_names = "sample_id",show_col_types = FALSE) %>% pull(sample_id)
sample_table <- read_csv("data/hpc/summary_data.csv",show_col_types = FALSE) %>% 
  select(sample_id,pop,location,reef) %>% 
  mutate(reef=ifelse(reef=="Outer","Offshore","Inshore"))

l1_covmat <- read_table("data/hpc/inshore_offshore/pcangsd_sig005/Sc0000013_chr1_RagTag.pcangsd.cov",col_names = F) %>% as.matrix()
l2_covmat <- read_table("data/hpc/inshore_offshore/pcangsd_sig005/Sc0000135_chr11_RagTag.pcangsd.cov",col_names = F) %>% as.matrix()
l3_covmat <- read_table("data/hpc/inshore_offshore/pcangsd_sig005/Sc0000214_Sc0000151_RagTag.pcangsd.cov",col_names = F) %>% as.matrix()
la_covmat <- read_table("data/hpc/inshore_offshore/pcangsd_sig005/Sc0000185_chr2a_RagTag.pcangsd.cov",col_names = F) %>% as.matrix()
lb_covmat <- read_table("data/hpc/inshore_offshore/pcangsd_sig005/Sc0000097_chr2b_RagTag.pcangsd.cov",col_names = F) %>% as.matrix()
```

```{r}
covmat2genotypes <- function(covmat,locus){
  colnames(covmat)<- sample_ids
  rownames(covmat) <- sample_ids
  pop_eigen <- eigen(covmat)

  pc1_loading <- data.frame(pc1=pop_eigen$vectors[,1],sample_id = sample_ids) %>% 
    left_join(sample_table,by = "sample_id") %>% 
    mutate(pop_order=site_order()[pop])

  pc1_genotypes <- pc1_loading %>% 
    mutate(cluster=kmeans(pc1,c(min(pc1),(min(pc1)+max(pc1))/2,max(pc1)))$cluster) %>%
    mutate(genotype = case_when(
      cluster==1 ~ "A/A",
      cluster==2 ~ "A/B",
      cluster==3 ~ "B/B"
    ))
  pc1_genotypes %>% mutate(locus=locus)
}

covmats <- list(l1_covmat,l2_covmat,l3_covmat,la_covmat,lb_covmat)
loci <- c("l1","l2","l3","la","lb")  

pc1_genotype_data <- map2_dfr(covmats,loci,covmat2genotypes)
```

To check that this automatic clustering method produces sensible cluster assignments we visualise asignments along with the original PC1 data for all inversion loci as in the plot below.

```{r}
  ggplot(pc1_genotype_data,aes(x=reorder(pop,pop_order),y=pc1)) + 
    geom_jitter(size=2,alpha=.8,width = 0.15,aes(color=as.factor(cluster))) +     
    theme_pubr(base_size = 10) + labs(y="PC1",x="Location", shape="", color="") + 
  theme(axis.text.x = element_text(angle=90), legend.position = "bottom") + 
  facet_wrap(~locus,scales = "free_y",ncol = 5) + 
  scale_color_discrete(name="Inversion Genotype",labels=c("A/A","A/B","B/B"))
```


```{r}
library(genetics)
library(adegenet)
library(hierfstat)

genotype_data <- pc1_genotype_data %>% 
  dplyr::select(-pc1,-cluster) %>% 
  pivot_wider(names_from = locus,values_from = genotype)

allele_data <- genotype_data %>% dplyr::select(l1,l2,l3,la,lb)
```

Having assigned genotypes we then performed two tests. First we tested to see if overall distribution of genotypes at each locus is in Hardy Weinberg Equilibrium (HWE). We found no significant deviations from HWE except for locus lb which had a p-value of 0.01. In this case the distinction between clusters is not particularly clear so the result might simply reflect bias in our cluster assignment.

```{r}
hwe_l1 <- HWE.test(genotype(allele_data$l1))
hwe_l2 <- HWE.test(genotype(allele_data$l2))
hwe_l3 <- HWE.test(genotype(allele_data$l3))
hwe_la <- HWE.test(genotype(allele_data$la))
hwe_lb <- HWE.test(genotype(allele_data$lb))
c("l1"=hwe_l1$test$p.value,"l2"=hwe_l2$test$p.value,"l3"=hwe_l3$test$p.value,"la"=hwe_la$test$p.value,"lb"=hwe_lb$test$p.value)
```

Next we calculated Fst per-locus between inshore and offshore.  There is no p-value calculation here, however, we can see that all values are low "<0.007" indicating no clear evidence for strong differential selection at these inversions.

```{r}
genind <- df2genind(allele_data,sep="/",ind.names = genotype_data$sample_id, pop = genotype_data$reef)

fst_l1 <- genet.dist(genind[loc="l1"], method = "WC84",diploid = TRUE)
fst_l2 <- genet.dist(genind[loc="l2"], method = "WC84",diploid = TRUE)
fst_l3 <- genet.dist(genind[loc="l3"], method = "WC84",diploid = TRUE)
fst_la <- genet.dist(genind[loc="la"], method = "WC84",diploid = TRUE)
fst_lb <- genet.dist(genind[loc="lb"], method = "WC84",diploid = TRUE)

c("l1"=fst_l1,"l2"=fst_l2,"l3"=fst_l3,"la"=fst_la,"lb"=fst_lb)
```





