GO term enrichment on genes implicated in outlier fst regions and
inversions
================

To investigate genes within inversions we first used bedtools to find
all genes that overlapped with inversions

``` bash
bedtools intersect -b <(bedtools sort -i inversions.bed) \
    -a aten_0.11.maker_post_001.genes.gff -wao | awk '$10!="."{OFS="\t";print $1,$2,$3,$4,$5,$6,$7,$8,$9,$13}' \
    > genes_in_inversions.tsv
```

This identified a total of 214 genes that overlapped with one of the
five inversion loci.

We then used `topGO` to identify GO terms that are enriched in this gene
set compared with the background (all genes in the genome). We used gene
functional annotation information obtained from BLAST searches against
the Swissprot database as described in @Cooke2020-qv

`topGO` was run using a nodeSize of 5 and the weight01 algorithm which
weights terms depending on position in the GO hierarchy to reduce the
chance that high level terms will be falsely called as significant due
to enrichment of more specific (lower) terms.

This revealed a single term “DNA Binding” that was

Although a total of 14 terms were significant at the p\<0.01 level many
of these were based on very few genes (3 or less). The most significant
term was DNA Binding (p=8.10E-05) and based on 14 genes.

Further investigation of the genes supporting enrichment of DNA binding
revealed that these genes were distributed across L1, L2 and L5
inversions.
